keyring/key.go

```diff
@@ -0,0 +1,168 @@
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+type Key interface {
+	JWASignatureAlgorithm() (jwa.SignatureAlgorithm, error)
+	JWKThumbprintSHA256() (string, error)
+	JWTSerializer() (*jwt.Serializer, error)
+	SSHFingerprintSHA256() (string, error)
+}
+
+type keyImpl struct {
+	comment string
+	jwk     jwk.Key
+}
+
+// Open reads a key from file, returning a Key
+func Open(path string) (Key, error) {
+	// Read the key file
+	keyData, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %v: %w", path, err)
+	}
+
+	return Parse(keyData)
+}
+
+// ParseString is a shortcut for calling Parse with a string
+func ParseString(s string) (Key, error) {
+	return Parse([]byte(s))
+}
+
+// Parse parsses raw key bytes
+func Parse(bytes []byte) (Key, error) {
+	// Try to parse the key as a JWK
+	if key, err := jwk.ParseKey(bytes); err == nil {
+		return &keyImpl{jwk: key}, nil
+	}
+
+	// Try to parse as PEM
+	if key, err := jwk.ParseKey(bytes, jwk.WithPEM(true)); err == nil {
+		return &keyImpl{jwk: key}, nil
+	}
+
+	// Try to get a raw crypto/* type from an OpenSSH private key
+	if key, err := ssh.ParseRawPrivateKey(bytes); err == nil {
+		// Load the raw crypto/* type
+		if converted, err := from(key); err == nil {
+			// Return the loaded key
+			return converted, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	return nil, errors.New("failed to parse as JWK, PEM, or OpenSSH private key")
+}
+
+// From creates a Key from go standard library key types
+func from(raw interface{}) (*keyImpl, error) {
+	// Handle other types more directly
+	switch raw := raw.(type) {
+	case *ed25519.PrivateKey:
+		if convertedToJWK, err := jwk.FromRaw(*raw); err == nil {
+			return &keyImpl{jwk: convertedToJWK}, nil
+		}
+	default:
+		if convertedToJWK, err := jwk.FromRaw(raw); err == nil {
+			return &keyImpl{jwk: convertedToJWK}, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to convert %T", raw)
+}
+
+// Raw returns the raw crypto/* type of the key
+func (k *keyImpl) Raw() (interface{}, error) {
+	// Convert the key to its raw key type from crypto/*
+	var rawKey interface{}
+	if err := k.jwk.Raw(&rawKey); err != nil {
+		return nil, fmt.Errorf("unable to convert to raw key: %w", err)
+	}
+
+	return rawKey, nil
+}
+
+// JWKThumbprintSHA256 returns the base64 encoded SHA256 JWK thumbprint of the key per RFC7638
+func (k *keyImpl) JWKThumbprintSHA256() (string, error) {
+	// Generate the SHA256 thumbprint
+	hash, err := k.jwk.Thumbprint(crypto.SHA256)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate thumbprint: %v", err)
+	}
+
+	// Return the base64 hash
+	return base64.RawURLEncoding.EncodeToString(hash), nil
+}
+
+// JWASignatureAlgorithm() returns a suitable signing algorithm for the key
+func (k *keyImpl) JWASignatureAlgorithm() (jwa.SignatureAlgorithm, error) {
+	// Inspect the key type
+	switch k.jwk.KeyType() {
+	case jwa.Ed25519:
```

Comment:
probably a simple mapping, 256-bit hashes suffice imo

Reply:
This relates more to the library than anything else, but why not pick the most secure option provided? Performance should still be zero concern, and if somebody wants more performance they should switch away from RSA rather than trying to improve within the scope of RSA.